### ISSUE TITLE
Write new stuff in Rust

### ISSUE THREAD

**Issue by selaux** [ja2-stracciatella/ja2-stracciatella#355]
I'm interested in Rust and just wanted to ask if we might switch to the rust language for new stuff. Rust is  interoperable with C/C++ and we get some of the improvements of the rust language:
- Memory safety
- An easier syntax
- Easy testing

But there are also drawbacks:
- Unfamiliar to some developers

But imho by using rust we could attract developers that are somehow detered by the current status of the codebase (a big ball of C-like C++). We could create a library (e.g. `libstracciatella`) that is completely written in rust, use it in C++ and link it statically/dynamically.

BTW: This could also be a step to a new license which all the rust code will live under (*GPL)

**Comment by lynxlynxlynx** [comment]
How dependency and setup heavy is that?

**Comment by selaux** [comment]
I think the only dependency in the beginning should be the rust compiler. Maybe I will create a small prototype where I migrate a single function or so, so we can see the impact.

**Comment by RinneT** [comment]
I'm quite interested in the idea. Had similar thoughts regarding Go, though it would be ill-suited - I think - and rust definitely sounds worth to learn.

Just thinking wether it might be a good idea to wait for the SDL2 branch to be merged, as that goes with large changes (I think).

**Comment by lynxlynxlynx** [comment]
We're not talking about a complete rewrite — that'd be mad.

**Comment by selaux** [comment]
It think even more important than SDL is the cmake PR, since we need to see how it integrates with the build system. Perspectively (my most favorite word) we could look at a complete rewrite, but since rust is mostly interoperable with C++ this can be done over months or rather years...

**Comment by derSteFfi** [comment]
Definitly interested here. I'm keen on the prototype...
A rewrite most of the time is a good way to find a lot of old bugs and straighten out a few things...

**Comment by lynxlynxlynx** [comment]
A rewrite most of the time is a good way to introduce a lot of new bugs. :P Usually regressions, as old complexity is confused by overcomplexity.

But in general, sure, the project is in flux. It started as a pure C app, then got partially plusplused, but there's still lots of cleanup to be done.

**Comment by selaux** [comment]
So I played around a little bit with rust and c++ and it's not as simple as I thought. Basically all memory that is allocated in rust needs to be freed by rust again. So returning anything else than simple types (ints, doubles, etc) gets pretty tedious.

A solution for that would to move entire "subsystems" to Rust so that C++ might query long living data structures which dont need to be deallocated "automatically". Or use something like the `AutoSGP*` structures that are currently there to automatically deallocate.

**Comment by Manuel-K** [comment]
I'm not quite sure how I should write this, but I don't think this was a very good idea.

While the Rust's syntax isn't that complicated, it's more complicated that it looks at the first glance. I've only read parts of the Rust book and while I should probably read more to get a better overview of the language, every time I've tried something seemingly simple I've spent a lot of time fighting against the compiler. The intention of memory safety in parallelized operations make some things that should be simple (like string manipulations) a lot more complicated.

On top of that the "standard library" is very small. Although that problem is partly mitigated by the huge number of crates they tend to bring other problems. They are in a constant state of flux (and they should be because they are a proving ground for new functionality) making working with them highly unstable. The only workaround is freezing a specific version, but that means that shipping an outdated version all the time, i.e. all bugs are preserved as well. In theory this problem could be solved by periodically checking for new versions of every fixed crate… not fun.

And even worse it complicates the build system because it introduces the need for a second compiler. On Arch Linux this is mostly unnoticeable, but it makes cross compilation situations and non standard build setups (e.g. msys2 on Windows #590) a lot harder.

And another thing: It makes debugging the code harder. At least in Qt Creator it's impossible to follow the function calls further than `RustInterface.h`.

The only thing Rust has going for our use case (embedding in a game) is that it is small. This is not a browser/server application/operation system. IMO we do not require the main features of Rust.

TL;DR: I'm not happy about the Rust integration.

OT: I think the name Rust is bad. In most situations you spend a lot of energy to avoid oxidation (i.e. rust) by adding oil, paint or other rust prevention methods.

**Comment by selaux** [comment]
@Manuel-K You might be surprised, but I agree with your point to some degree, although I don't share all of your opinions. Basically I agree that the current integration of Rust code into the C++ code base is not sustainable for the long term. Basically there is too much boilerplate included and the resulting dev experience is still horrible (calling c-style functions, no proper debugging possible, very cryptic errors in the ffi layer). 

About your points:

- Fighting against the compiler: Yes that is something you do quite often in the beginning, but the only thing Rust does is trying to help you 😉 . It is true that rust, although being a young language, has a bunch of syntax and concepts to learn. On the other hand a newcomer to C++ would argue the same and C++ often wont give the users hints to some of the issues that the Rust compiler detects.
- The standard library is very small: Cant share this one. I'm not missing a lot in what `std` provides and it works the same on all platforms 😉 . Also as you already pointed out there are crates, there is actual dependency management provided by the language / build system. Looking at how we manage dependencies right now, this is a huge advantage of Rust. The disadvantages you mention are very easily mitigated (Cargo.lock in combination with update bot).
- Build system: I agree. Of course adding a new compiler and language will make the build system more complicated. This is something rust developers are trying to make easier (https://github.com/rust-lang/rust-roadmap/issues/12). I would also argue that it is generally hard to build a C++ build system that runs on as many build configurations as we are trying to support and I don't understand why we are trying to support so many.

What I think it would take to make rust integration good:
- Better build system integration: Build as `ExternalProject` find with `find_package` in order to mitigate differing shared library names on different systems.
- Make it a wrapper around "proper" rust code: Split rust code into modules, so it is structured more. Write Rust code with less of a "I have to drop to C types at some point"-attitude. Then have automated C++ bindings written for the "properly written" rust code. I imagine this being a ton of work as there is no C++ wrapper generator I know of out there.
- Make it debuggable
- Learn Rust: Easy for me to say 😀 

So I think that we need either do some investing to make Rust integration sustainable, or drop it again. As there is not a lot of Rust code present yet, this should be possible, but also some work.

TLDR: There are advantages and disadvantages of course and we need to decide wether to keep or drop Rust.

**Comment by Manuel-K** [comment]
It would not be that much work to remove the Rust code:

We can revive the config reading/writing code that uses RapidJSON from here: https://github.com/ja2-stracciatella/ja2-stracciatella/pull/548/commits/68475157636d0f3325f30b5c44bb2dd528db7305

We need the path functions for the launcher and the arg parser.

I want to remove the boost::filesystem dependency as well. That requires rewriting some of the path functions. Most of the stuff is very ease if we only require relative paths. Getting absolute paths is annoying, because we would need at least `GetFullPathNameW` for Windows and `canonicalize_file_name` for Linux… I'll create a separate issue about this.

I think integrating the arg parser and the config reader/writer into a `Config` class that can get its settings from both sources would be a good idea.
